File: python/sorting/bubble.py
def bubble_sort(arr):
    """ O(n^2): Return a sorted array """
    n = len(arr)
    for i in range(n):
        for j in range(n-1):
            if arr[j] > arr[j+1]:
                arr[j+1], arr[j] = arr[j], arr[j+1] # swap values
    return arr

if __name__ == "__main__":
    print(bubble_sort([3, -2, 9, 0, 0, 12, -5, 8]))

File: src/sorting/bubble.rs
pub fn bubble_sort<T: Ord + Copy>(arr: &mut [T]) {
    let len: usize = arr.len();
    for _ in 0..len {
        for i in 0..len-1 {
            if arr[i] > arr[i+1] {
                arr.swap(i, i+1);
            }
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn bubble_sort_test() {
        let mut arr: Vec<i32> = vec![3, -2, 9, 0, 12, -5, 8, 0];
        bubble_sort(&mut arr);
        assert_eq!(arr, vec![-5, -2, 0, 0, 3, 8, 9, 12]);
    }
}